tickets/DM-46157: Move Nublado Client into Nublado monorepo (#401)
lsst-sqre/nublado: docs/client/index.rst

############
Client guide
############

This page describes the use of the ``rubin.nublado.client.NubladoClient`` class and its provided testing classes.

Installing the Client
=====================

The client can be installed from Pypi with:

.. prompt:: bash

    pip install rubin-nublado-client

.. _client-usage:

Using the Client
================

> The ``NubladoClient`` is designed to make interaction with JuptyerHub and Jupyterlab, as they are configured in the RSP environment, easy.

Reviewer: ```suggestion
The ``NubladoClient`` is designed to make interaction with JupyterHub and Jupyterlab, as they are configured in the RSP environment, easy.
```

Sequence of Events
==================

A typical interaction with the client usually looks like this:

#. Authenticate to the Hub with the ``auth_to_hub()`` method.
#. Determine whether or not you have a running lab with ``is_lab_stopped()``.
#. If you need to, spawn a lab with ``spawn_lab()``.
#. Wait for the lab to spawn by looping through ``watch_spawn_progress()`` until you get a progress message indicating the lab is ready.
#. Authenticate to the Lab with ``auth_to_lab()``
#. Use ``open_lab_session()`` as a context manager to interact with your lab.
#. Do whatever it is you wanted to do with the lab; more below.
#. When done, use ``stop_lab()`` to shut down the lab, if desired.

The ``client_test`` test in the client test suite steps through this process and may be a useful model.

.. _lab-interaction:

Interacting With The Lab
========================

``NubladoClient`` provides three methods of interacting with a spawned lab.  These are methods on the ``JupyterLabSession`` object you will have available inside the session context manager.  They are:

> #. ``run_python()``.  This runs a string representing arbitrary Python code and returns streamed results (which is to say, ``stdout`` wihin each cell).  Of course, this code will run in the Lab environment specified by the session, which will be bound to a particuar kernel (usually ``LSST`` within the RSP), and will have access to whatever is installed in that kernel.

Reviewer: ```suggestion
#. ``run_python()``.  This runs a string representing arbitrary Python code and returns streamed results (which is to say, ``stdout`` within each cell).  Of course, this code will run in the Lab environment specified by the session, which will be bound to a particular kernel (usually ``LSST`` within the RSP), and will have access to whatever is installed in that kernel.
```

Examples
--------

The following uses the client to determine whether a user notebook is
running and to start it if necessary.  It does not run any code:

.. code-block:: python

    """Ensure that the user has a running lab."""
    
    import asyncio
    import structlog
    from rubin.nublado.client import NubladoClient
    from rubin.nublado.client.models import (
	NubladoImage,
	NubladoImageByClass,
	NubladoImageClass,
	NubladoImageSize,
    )
    
    LAB_SPAWN_TIMEOUT = 90

    async def ensure_lab() -> None:
        """Start a Lab if one is not present."""
        client = NubladoClient(
 	    user=User(username="some-user", token="some-token"),
            base_url="https://data.example.org"
        )
        await client.auth_to_hub()
        stopped = await client.is_lab_stopped()
	if stopped:

> 	    image=NubladoImageByClass(

Reviewer: Looks like some tabs that need to be spaces here

                image_class=NubladoImageClass.RECOMMENDED,
                size=NubladoImageSize.Medium
            )
            await client.spawn_lab(image)
            progress = client.watch_spawn_progress()
            async with asyncio.timeout(LAB_SPAWN_TIMEOUT):
                async for message in progress:
                    if message.ready:
                        break

    asyncio.run(ensure_lab)

The next example assumes that you have already done the above--that is, you know the user already has a running Lab--and that you had a burning desire to have an RSP instance calculate 2+2 for you.

.. _Mocks:

Mocks
=====

In the module ``rubin.nublado.client.testing`` you will find the ``MockJupyter`` class.
This provides a simulation of the RSP Nublado Hub/Proxy/Controller environment, as well as a partial simulation of the Labs it spawns.
The reason you would use this is to be able to meaningfully test your service without having to test against a live RSP or spin up your own RSP to test the service against.

> Although there are quite a few additional classes within the module, ``MockJupyter`` should be the only one you need directly, except to set up the test fixure.

Reviewer: ```suggestion
Although there are quite a few additional classes within the module, ``MockJupyter`` should be the only one you need directly, except to set up the test fixture.
```
